### PR TITLE
support prepare and onPrepare for RxAudioPlayer

### PR DESCRIPTION
**Pull request by sajadshokri**
and also pause and resume song

### CODE REVIEW

File: rxandroidaudio/src/main/java/com/github/piasy/rxandroidaudio/RxAudioPlayer.java [hunk 1]
```diff
@@ -78,19 +80,19 @@ public int progress() {
...
      */
-    public Single<Boolean> play(@NonNull final PlayConfig config) {
+    public Observable<Boolean> play(@NonNull final PlayConfig config) {
         if (config.mType == PlayConfig.TYPE_FILE && config.mAudioFile != null &&
                 config.mAudioFile.exists()) {
-            return Single.create(new Single.OnSubscribe<Boolean>() {
+            return Observable.create(new Observable.OnSubscribe<Boolean>() {
                 @Override
-                public void call(final SingleSubscriber<? super Boolean> singleSubscriber) {
+                public void call(final Subscriber<? super Boolean> singleSubscriber) {
```
Comment (Piasy) [inline rxandroidaudio/src/main/java/com/github/piasy/rxandroidaudio/RxAudioPlayer.java]: Now that it's not `SingleSubscriber` any more, we should not name it `singleSubscriber`.

File: rxandroidaudio/src/main/java/com/github/piasy/rxandroidaudio/RxAudioPlayer.java [hunk 2]
```diff
@@ -131,11 +135,75 @@ public void call(final SingleSubscriber<? super Boolean> singleSubscriber) {
...
+                    Log.d(TAG, "MediaPlayer to start play: " + config.mAudioFile.getName());
+                    mPlayer = new MediaPlayer();
+                    try {
+                        mPlayer.setDataSource(config.mAudioFile.getAbsolutePath());
+                        setMediaPlayerListener(singleSubscriber);
+                        mPlayer.setVolume(config.mLeftVolume, config.mRightVolume);
+                        mPlayer.setLooping(config.mLooping);
+                        mPlayer.prepare();
+                        singleSubscriber.onNext(true);
+//                        mPlayer.start();
```
Comment (Piasy) [inline rxandroidaudio/src/main/java/com/github/piasy/rxandroidaudio/RxAudioPlayer.java]: You could remove this line.

File: rxandroidaudio/src/main/java/com/github/piasy/rxandroidaudio/RxAudioPlayer.java [hunk 3]
```diff
@@ -131,11 +135,75 @@ public void call(final SingleSubscriber<? super Boolean> singleSubscriber) {
...
+                    stopPlay();
+
+                    Log.d(TAG, "MediaPlayer to start play: " + config.mAudioResource);
+                    mPlayer = MediaPlayer.create(config.mContext, config.mAudioResource);
+                    try {
+                        setMediaPlayerListener(singleSubscriber);
+                        mPlayer.setVolume(config.mLeftVolume, config.mRightVolume);
+                        mPlayer.setLooping(config.mLooping);
+                        singleSubscriber.onNext(true);
+//                        mPlayer.start();
```
Comment (Piasy) [inline rxandroidaudio/src/main/java/com/github/piasy/rxandroidaudio/RxAudioPlayer.java]: Remove this line.

File: rxandroidaudio/src/main/java/com/github/piasy/rxandroidaudio/RxAudioPlayer.java [hunk 4]
```diff
@@ -148,7 +216,7 @@ public void onCompletion(MediaPlayer mp) {
                     @Override
                     public void call(Long aLong) {
                         stopPlay();
-                        singleSubscriber.onSuccess(true);
+                        singleSubscriber.onCompleted(/*true*/);
```
Comment (Piasy) [inline rxandroidaudio/src/main/java/com/github/piasy/rxandroidaudio/RxAudioPlayer.java]: Name it as `subscriber`, and remove `/*true*/`.

File: rxandroidaudio/src/main/java/com/github/piasy/rxandroidaudio/RxAudioPlayer.java [hunk 5]
```diff
@@ -131,11 +135,75 @@ public void call(final SingleSubscriber<? super Boolean> singleSubscriber) {
...
         } else {
-            return Single.error(new IllegalArgumentException(""));
+            return Observable.error(new IllegalArgumentException(""));
+        }
+    }
+
+    /**
+     * prepare audio from local file. should be scheduled in IO thread.
+     */
+    public Observable<Boolean> prepare(@NonNull final PlayConfig config) {
```
Comment (Piasy) [inline rxandroidaudio/src/main/java/com/github/piasy/rxandroidaudio/RxAudioPlayer.java:145]: After we get prepared via `prepare`, how could we play it? and when would the returned `Observable` complete?

Comment (sajadshokri) [inline rxandroidaudio/src/main/java/com/github/piasy/rxandroidaudio/RxAudioPlayer.java:145]: after preparing `onNext` will be called and we can call `resume` to play the audio. and when the audio is finished `complete` will be called.

```
 mRxAudioPlayer.prepare(playConfig)
            .subscribeOn(Schedulers.io())
            .subscribe(new Action1<Boolean>() {
                @Override
                public void call(Boolean aBoolean) {//prepared
                    mRxAudioPlayer.resume();// play
                }
            }, new Action1<Throwable>() {
                @Override
                public void call(Throwable throwable) {//Error

                }
            }, new Action0() {
                @Override
                public void call() {
                    //end of audio
                }

                );
```

Comment (Piasy) [inline rxandroidaudio/src/main/java/com/github/piasy/rxandroidaudio/RxAudioPlayer.java:145]: Got it :)

**Comment by Piasy** [comment]
@sajadshokri `v1.3.0` has been released.

**Pull request merged**